src/protocol/modulus_conversion/mod.rs

```diff
@@ -66,14 +66,14 @@ pub fn split_into_multi_bit_slices<T: Clone>(
 ///     ...
 ///     `[ row[n].bit0, row[n].bit1, ..., row[n].bit31 ]`,
 ///  `]`
-pub fn combine_slices<T: Clone>(
-    input: &[Vec<Vec<T>>],
+pub fn combine_slices<'a, T: Clone + 'a>(
```

Comment:
This looks like the much-dreaded `transpose()` we tried very hard to eliminate.

Though this takes and produces an iterator, it is still very expensive because it allocates a vector.

The input vector appears to be a 3d collection, but it appears as though that is only a fraction of `num_bits` item deep on the third dimension.  (The documentation here is unhelpful as both input and output in the comment are 2d.  It would be easier to understand if this explained some of the context into which this function is expected to fit.)

It seems to me like this is a hard problem to solve if you are operating on an entire collection in this way, as the chunks into which we split the bits are spread throughout the input.  To avoid the allocation, you need all bits that need to be adjacent in the output to be adjacent in the input.

For that to work out, we probably need to save something, but I would prefer it if we changed the bit conversion so that we store all bits for each record in a single row, rather than produce a 3d vector.  We then need to index into the correct bits as needed when sorting, but at least we don't need this expensive conversion once the bits are sorted.

Reply:
The documentation was incorrect, I should have added one [] in each row.

Reply:
In the next PR i was hoping that we could return iterator from convert_all_bits which would be slices based on what sort needs in that iteration (e.g. iterator on m..n bits for each record). By doing what you suggested, we would need to pre-process modulus conversion on all the records together. We also have an upcoming optimization to interleave modulus conversion with sort which would be easier with the current format: https://github.com/private-attribution/ipa/blob/main/src/protocol/sort/generate_permutation_opt.rs#L172

Let me know your thoughts!

Reply:
The problem with this is that you are doing a full-sized allocation here, which is maybe OK if this is the single place where that allocation occurs, but that doesn't seem to be where we are.

Reply:
Agree this PR does not take care of this. It is just dumbly replacing Vec's by iterator. Let me remove combine_slices change from this PR. This issue should be handled properly once XorReplicated is refactored. Sort should then convert XorReplicated shares, use them to sort, combine and return it to the caller. Does that sound ok?